app/Http/Controllers/AlbumController.php

```diff
@@ -5,344 +5,308 @@
 	{
-		$request->validate([
-			'license' => ['required', 'string', Rule::in(Helpers::get_all_licenses())],
-		]);
-		$setLicense->do($request['albumID'], $request['license']);
-
-		return response()->noContent();
+		$request->album()->license = $request->license();
+		$request->album()->save();
 	}
 
 	/**
-	 * Delete the album and all pictures in the album.
+	 * Delete the album and all of its pictures.
+	 *
+	 * @param DeleteAlbumsRequest $request
 	 *
-	 * @param AlbumIDsRequest $request
-	 * @param Delete          $delete
+	 * @return void
 	 *
-	 * @return IlluminateResponse
+	 * @throws LycheeException
 	 */
-	public function delete(AlbumIDsRequest $request, Delete $delete): IlluminateResponse
+	public function delete(DeleteAlbumsRequest $request): void
 	{
-		if ($delete->do(explode(',', $request['albumIDs']))) {
-			return response()->noContent();
-		} else {
-			return response('', 500);
+		/** @var AbstractAlbum $album */
+		foreach ($request->albums() as $album) {
+			if ($album instanceof BaseAlbum || $album instanceof UnsortedAlbum) {
+				$album->delete();
+			}
 		}
 	}
 
 	/**
 	 * Merge albums. The first of the list is the destination of the merge.
 	 *
-	 * @param AlbumIDsRequest $request
-	 * @param Merge           $merge
+	 * @param MergeAlbumsRequest $request
+	 * @param Merge              $merge
 	 *
-	 * @return IlluminateResponse
+	 * @return void
+	 *
+	 * @throws LycheeException
+	 * @throws ModelNotFoundException
 	 */
-	public function merge(AlbumIDsRequest $request, Merge $merge): IlluminateResponse
+	public function merge(MergeAlbumsRequest $request, Merge $merge): void
 	{
-		$targetAlbumID = $request['albumID'];
-		$albumIDs = explode(',', $request['albumIDs']);
-		$merge->do($targetAlbumID, $albumIDs);
-
-		return response()->noContent();
+		$merge->do($request->album(), $request->albums());
 	}
 
 	/**
 	 * Move multiple albums into another album.
 	 *
-	 * @param AlbumIDsRequest $request
-	 * @param Move            $move
+	 * @param MoveAlbumsRequest $request
+	 * @param Move              $move
 	 *
-	 * @return IlluminateResponse
+	 * @return void
+	 *
+	 * @throws LycheeException
+	 * @throws ModelNotFoundException
 	 */
-	public function move(AlbumIDsRequest $request, Move $move): IlluminateResponse
+	public function move(MoveAlbumsRequest $request, Move $move): void
 	{
-		$targetAlbumID = $request['albumID'];
-		$albumIDs = explode(',', $request['albumIDs']);
-		$move->do($targetAlbumID, $albumIDs);
-
-		return response()->noContent();
+		$move->do($request->album(), $request->albums());
 	}
 
 	/**
 	 * Set if an album contains sensitive pictures.
 	 *
-	 * @param AlbumIDRequest $request
-	 * @param SetNSFW        $setNSFW
+	 * This method is either a misnomer and should rather be called
+	 * `toggleNSFW` or the request `SetAlbumNSFWRequest` should contain an
+	 * explicit boolean with the new NSFW state.
+	 *
+	 * TODO: Fix the method, see above.
```

Comment:
Just my $0.02: I don't like toggle methods, and especially when there's an unreliable network between the client and the server. I vote for adding a boolean argument.

Reply:
Totally agreed, but that the NSFW attribute is being toggled instead of set explicitly has not been invented by me.

I fixed the that problem for `setPublic` of a single photo which suffered from the same problem.

**Current situation on master for "setPublic" (note that `setPublic` actually means "toggle"):**

https://github.com/LycheeOrg/Lychee/blob/05f21d48bd7205016c7b8618da711c4d7cf4d959/app/Actions/Photo/Toggle.php#L25-L30

and

https://github.com/LycheeOrg/Lychee/blob/05f21d48bd7205016c7b8618da711c4d7cf4d959/app/Actions/Photo/SetPublic.php#L5-L11

**New situation for "setPublic"**

https://github.com/LycheeOrg/Lychee/blob/89a948323c786c0bc41bcddd2c9c3a15cda9ea3f/app/Http/Controllers/PhotoController.php#L172-L176

**Problems**

It crossed my mind just now that I could have done the same for `Albums::setNSFW` because it only affects a single instance of an album. However, we also have some `set...`-methods, like `toggleStar` which toggles but for _multiple photos_. See here:

https://github.com/LycheeOrg/Lychee-front/blob/d7f12fb3ccf94133c41111b47b14bea4c644e51e/scripts/main/photo.js#L526-L540

I wonder what we should do with that. I never have tested it myself with the frontend what happens if one selects multiple photos and clicks the star button. Is this supported at all for multiple photos?

Reply:
> what happens if one selects multiple photos and clicks the star button.

This will not happen, the star button is only available in photo header, not in album header if I remember correctly.

Reply:
But the action is also available from the context menu. You can select multiple photos in the album view, do a right click and then select "Mark as favorites". The action is actually also a misnomer, because you don't unconditionally mark the photos as favorites, but the status is toggle.

I just tried it. I even selected a mix of starred and unstarred photos and after the action the stars are inverted. However, I wonder if this is what _should_ happen. IMHO, this function is awkward.

Reply:
Definitively NOT the expected behavior. :')

Reply:
Which one?

Reply:
Done in https://github.com/LycheeOrg/Lychee/pull/1217/commits/15374daa8b425704da2385549ac0b3ee6be2f98c and in https://github.com/LycheeOrg/Lychee-front/pull/282/commits/e712c60761d6bcc5428e34f729f2974262fa04fe.

Reply:
I had only a brief look at the commit and I like most of it but I'm not sure about what looks like adding a separate "Set" and "Unset" options to the multi-select context menu. Honestly, I think we should completely remove starring in that case (so, your second alternative above). We could also leave a (properly labeled) "Toggle Star Selected" in case all the selected photos are either starred or unstarred -- or, if we wanted to be _real_ fancy, display it as either "Star Selected" or "Unstar Selected" depending on the condition of the selected photos. While doing that, I would also get rid of "Rename Selected" -- it strikes me as completely pointless. But we can do that after this PR, of course.

Reply:
> Honestly, I think we should completely remove starring in that case (so, your second alternative above)

That would also have been my first choice, but I feared that some users might start yelling, because we removed a feature. But it would be so much more consistent with other binary attributes like NSFW, etc.

> We could also leave a (properly labeled) "Toggle Star Selected" in case all the selected photos are either starred or unstarred -- or, if we wanted to be real fancy, display it as either "Star Selected" or "Unstar Selected" depending on the condition of the selected photos.

That was what I wanted to implement as an alternative to my first choice above until I noticed that this isn't possible efficiently. The main obstacle is that in the album view we we don't have any efficient access to the selected photos (we only have the photos IDs) and thus we cannot _efficiently_ determine wether all are starred or unstarred. Hence, we can only submit a request to star/unstar the set of photos belonging to the IDs without looking at their current state.

That is why I decided to add both options. I don't like it either. I would rather prefer to use the first solution.

I assume that this is also the reason why the previous and unfortunate option "Toggle star" made it. It simply asks the sever to toggle the star attribute of a set of photos with a possibly mixed set of attributes without being aware of it. An old comment in the code said "ugly, but efficient".

Reply:
I suggest that we leave it as-is for now and when I get to the front end PR and then to testing I'll see if I can figure out a way to improve it.

Reply:
I pushed a change to the front end PR to implement the selective Star/Unstar. Sure, using `album.getByID` for each element of an array is not very efficient, but I think it's acceptable, especially in a GUI context (where a human is likely to be the slowest link).